File: packages/ios/Sources/Components/Combobox/Combobox.swift
import SwiftUI

public struct KozmosCombobox: View {
    @Binding public var value: String
    @Binding public var inputValue: String
    public let options: [KozmosListboxOption]
    public let label: String?
    public let placeholder: String
    public let disabled: Bool
    public let readOnly: Bool
    public let status: KozmosInputStatus
    public let error: Bool
    public let helperText: String?
    public let errorMessage: String?
    public let emptyText: String
    public let clearable: Bool

    @State private var isOpen: Bool

    public init(
        value: Binding<String> = .constant(""),
        inputValue: Binding<String> = .constant(""),
        options: [KozmosListboxOption],
        label: String? = nil,
        placeholder: String = "Select option",
        disabled: Bool = false,
        readOnly: Bool = false,
        status: KozmosInputStatus = .default,
        error: Bool = false,
        helperText: String? = nil,
        errorMessage: String? = nil,
        emptyText: String = "No results found",
        clearable: Bool = true,
        defaultOpen: Bool = false
    ) {
        self._value = value
        self._inputValue = inputValue
        self.options = options
        self.label = label
        self.placeholder = placeholder
        self.disabled = disabled
        self.readOnly = readOnly
        self.status = status
        self.error = error
        self.helperText = helperText
        self.errorMessage = errorMessage
        self.emptyText = emptyText
        self.clearable = clearable
        self._isOpen = State(initialValue: defaultOpen)
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing75) {
            if let label = label {
                Text(label)
                    .font(.subheadline.weight(.semibold))
                    .foregroundColor(labelColor)
            }

            HStack(spacing: 0) {
                TextField(placeholder, text: $inputValue)
                    .disabled(disabled || readOnly)
                    .font(.subheadline)
                    .foregroundColor(textColor)
                    .padding(.leading, KozmosDimensions.primitivesLayoutSpacing150)
                    .frame(height: 44)
                    .onTapGesture {
                        if !disabled && !readOnly { isOpen = true }
                    }
                    .onChange(of: inputValue) { _ in
                        if !disabled && !readOnly { isOpen = true }
                    }

                if clearable && !inputValue.isEmpty && !disabled && !readOnly {
                    Button(action: clearSelection) {
                        Image(systemName: "xmark")
                            .font(.subheadline.weight(.semibold))
                            .frame(width: 36, height: 44)
                    }
                    .buttonStyle(.plain)
                    .foregroundColor(KozmosColors.primitivesColorsForeground500)
                    .accessibilityLabel("Clear selection")
                }

                Button(action: toggleOpen) {
                    Image(systemName: "chevron.down")
                        .font(.subheadline.weight(.semibold))
                        .rotationEffect(.degrees(isOpen ? 180 : 0))
                        .frame(width: 44, height: 44)
                }
                .buttonStyle(.plain)
                .disabled(disabled || readOnly)
                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                .accessibilityLabel(isOpen ? "Close options" : "Open options")
            }
            .frame(maxWidth: .infinity, minHeight: 44, maxHeight: 44)
            .background(fieldBackgroundColor)
            .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100))
            .overlay(
                RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100)
                    .stroke(fieldBorderColor, lineWidth: 1)
            )

            if isOpen {
                if filteredOptions.isEmpty {
                    Text(emptyText)
                        .font(.subheadline)
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)
                        .padding(KozmosDimensions.primitivesLayoutSpacing150)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .background(KozmosColors.primitivesColorsBackground0)
                        .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100))
                        .overlay(
                            RoundedRectangle(cornerRadius: KozmosDimensions.primitivesLayoutRadius100)
                                .stroke(KozmosColors.primitivesColorsForeground500, lineWidth: 1)
                        )
                } else {
                    KozmosListbox(
                        options: filteredOptions,
                        selectedValues: listboxSelection,
                        multiple: false,
                        disabled: disabled,
                        maxHeight: 256
                    ) { nextValues, option in
                        value = nextValues.first ?? ""
                        inputValue = option.label
                        isOpen = false
                    }
                }
            }

            if let supportingText = supportingText, !supportingText.isEmpty {
                Text(supportingText)
                    .font(.subheadline)
                    .foregroundColor(supportingTextColor)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var filteredOptions: [KozmosListboxOption] {
        let query = inputValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !query.isEmpty else { return options }
        return options.filter { option in
            option.label.lowercased().contains(query)
                || option.value.lowercased().contains(query)
                || (option.description?.lowercased().contains(query) ?? false)
        }
    }

    private var listboxSelection: Binding<[String]> {
        Binding(get: { value.isEmpty ? [] : [value] }, set: { value = $0.first ?? "" })
    }

    private func toggleOpen() {
        guard !disabled, !readOnly else { return }
        isOpen.toggle()
    }

    private func clearSelection() {
        value = ""
        inputValue = ""
        isOpen = false
    }

    private var effectiveStatus: KozmosInputStatus {
        error ? .error : status
    }

    private var supportingText: String? {
        errorMessage ?? helperText
    }

    private var fieldBackgroundColor: Color {
        (disabled || readOnly) ? KozmosColors.primitivesColorsBackground100 : KozmosColors.primitivesColorsBackground0
    }

    private var fieldBorderColor: Color {
        switch effectiveStatus {
        case .error:
            return KozmosColors.primitivesColorsEmotionalDanger600
        case .warning:
            return KozmosColors.primitivesColorsEmotionalAlert600
        case .success:
            return KozmosColors.primitivesColorsEmotionalSuccess600
        case .default:
            return KozmosColors.primitivesColorsForeground500
        }
    }

    private var labelColor: Color {
        if disabled { return KozmosColors.primitivesColorsForeground500 }
        switch effectiveStatus {
        case .error:
            return KozmosColors.primitivesColorsEmotionalDanger600
        case .warning:
            return KozmosColors.primitivesColorsEmotionalAlert600
        case .success:
            return KozmosColors.primitivesColorsEmotionalSuccess600
        case .default:
            return KozmosColors.primitivesColorsForeground100
        }
    }

    private var supportingTextColor: Color {
        if disabled { return KozmosColors.primitivesColorsForeground500 }
        switch effectiveStatus {
        case .error:
            return KozmosColors.primitivesColorsEmotionalDanger600
        case .warning:
            return KozmosColors.primitivesColorsEmotionalAlert600
        case .success:
            return KozmosColors.primitivesColorsEmotionalSuccess600
        case .default:
            return KozmosColors.primitivesColorsForeground500
        }
    }

    private var textColor: Color {
        if disabled { return KozmosColors.primitivesColorsForeground500 }
        if effectiveStatus == .error { return KozmosColors.primitivesColorsEmotionalDanger600 }
        return KozmosColors.primitivesColorsForeground0
    }
}

File: packages/android/src/main/java/com/kozmos/components/Combobox/Combobox.kt
package com.kozmos.components.combobox

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.dp
import com.kozmos.components.input.KozmosInputStatus
import com.kozmos.components.listbox.KozmosListbox
import com.kozmos.components.listbox.KozmosListboxOption
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

@Composable
fun KozmosCombobox(
    value: String,
    onValueChange: (String, KozmosListboxOption?) -> Unit,
    inputValue: String,
    onInputValueChange: (String) -> Unit,
    options: List<KozmosListboxOption>,
    modifier: Modifier = Modifier,
    label: String? = null,
    placeholder: String = "Select option",
    enabled: Boolean = true,
    readOnly: Boolean = false,
    status: KozmosInputStatus = KozmosInputStatus.Default,
    error: Boolean = false,
    helperText: String? = null,
    errorMessage: String? = null,
    emptyText: String = "No results found",
    clearable: Boolean = true,
    expanded: Boolean? = null,
    defaultExpanded: Boolean = false,
    onExpandedChange: ((Boolean) -> Unit)? = null
) {
    var internalExpanded by rememberSaveable { mutableStateOf(defaultExpanded) }
    val isExpanded = expanded ?: internalExpanded
    val setExpanded: (Boolean) -> Unit = { next ->
        if (expanded == null) internalExpanded = next
        onExpandedChange?.invoke(next)
    }
    val effectiveStatus = if (error) KozmosInputStatus.Error else status
    val supportingText = errorMessage ?: helperText
    val filteredOptions = options.filter { option ->
        val query = inputValue.trim().lowercase()
        query.isEmpty() ||
            option.label.lowercase().contains(query) ||
            option.value.lowercase().contains(query) ||
            (option.description?.lowercase()?.contains(query) == true)
    }
    val colors = selectionFieldColors(effectiveStatus, enabled, readOnly)
    val fieldShape = RoundedCornerShape(KozmosDimensions.primitivesLayoutRadius100)

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing75)
    ) {
        if (label != null) {
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = colors.label
            )
        }

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .height(44.dp)
                .background(colors.background, fieldShape)
                .border(1.dp, colors.border, fieldShape),
            verticalAlignment = Alignment.CenterVertically
        ) {
            BasicTextField(
                value = inputValue,
                onValueChange = {
                    onInputValueChange(it)
                    if (enabled && !readOnly) setExpanded(true)
                },
                enabled = enabled,
                readOnly = readOnly,
                singleLine = true,
                textStyle = MaterialTheme.typography.bodyMedium.copy(color = colors.text),
                cursorBrush = SolidColor(KozmosThemeTokens.primitivesColorsTheme500),
                modifier = Modifier
                    .weight(1f)
                    .fillMaxHeight(),
                decorationBox = { innerTextField ->
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .fillMaxHeight()
                            .padding(start = KozmosDimensions.primitivesLayoutSpacing150),
                        contentAlignment = Alignment.CenterStart
                    ) {
                        if (inputValue.isEmpty()) {
                            Text(
                                text = placeholder,
                                style = MaterialTheme.typography.bodyMedium,
                                color = colors.placeholder
                            )
                        }
                        innerTextField()
                    }
                }
            )

            if (clearable && inputValue.isNotEmpty() && enabled && !readOnly) {
                IconButton(
                    onClick = {
                        onValueChange("", null)
                        onInputValueChange("")
                        setExpanded(false)
                    },
                    modifier = Modifier.size(36.dp),
                    colors = IconButtonDefaults.iconButtonColors(contentColor = KozmosThemeTokens.primitivesColorsForeground500)
                ) {
                    Icon(Icons.Default.Close, contentDescription = "Clear selection", modifier = Modifier.size(16.dp))
                }
            }

            IconButton(
                onClick = { if (enabled && !readOnly) setExpanded(!isExpanded) },
                enabled = enabled && !readOnly,
                modifier = Modifier.size(44.dp),
                colors = IconButtonDefaults.iconButtonColors(
                    contentColor = KozmosThemeTokens.primitivesColorsForeground500,
                    disabledContentColor = KozmosThemeTokens.primitivesColorsForeground500
                )
            ) {
                Icon(
                    imageVector = Icons.Default.KeyboardArrowDown,
                    contentDescription = if (isExpanded) "Close options" else "Open options",
                    modifier = Modifier
                        .size(20.dp)
                        .rotate(if (isExpanded) 180f else 0f)
                )
            }
        }

        if (isExpanded) {
            if (filteredOptions.isEmpty()) {
                Text(
                    text = emptyText,
                    style = MaterialTheme.typography.bodyMedium,
                    color = KozmosThemeTokens.primitivesColorsForeground500,
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(KozmosThemeTokens.primitivesColorsBackground0, fieldShape)
                        .border(1.dp, KozmosThemeTokens.primitivesColorsForeground500, fieldShape)
                        .padding(KozmosDimensions.primitivesLayoutSpacing150)
                )
            } else {
                KozmosListbox(
                    options = filteredOptions,
                    selectedValues = if (value.isEmpty()) emptyList() else listOf(value),
                    onSelectionChange = { nextValues, option ->
                        onValueChange(nextValues.firstOrNull().orEmpty(), option)
                        onInputValueChange(option.label)
                        setExpanded(false)
                    },
                    enabled = enabled,
                    multiple = false
                )
            }
        }

        if (!supportingText.isNullOrBlank()) {
            Text(
                text = supportingText,
                style = MaterialTheme.typography.bodyMedium,
                color = colors.helper
            )
        }
    }
}

internal data class SelectionFieldColors(
    val background: androidx.compose.ui.graphics.Color,
    val border: androidx.compose.ui.graphics.Color,
    val text: androidx.compose.ui.graphics.Color,
    val placeholder: androidx.compose.ui.graphics.Color,
    val label: androidx.compose.ui.graphics.Color,
    val helper: androidx.compose.ui.graphics.Color
)

@Composable
internal fun selectionFieldColors(
    status: KozmosInputStatus,
    enabled: Boolean,
    readOnly: Boolean
): SelectionFieldColors {
    val statusColor = when (status) {
        KozmosInputStatus.Error -> KozmosThemeTokens.primitivesColorsEmotionalDanger600
        KozmosInputStatus.Warning -> KozmosThemeTokens.primitivesColorsEmotionalAlert600
        KozmosInputStatus.Success -> KozmosThemeTokens.primitivesColorsEmotionalSuccess600
        KozmosInputStatus.Default -> null
    }
    val textColor = when {
        status == KozmosInputStatus.Error -> KozmosThemeTokens.primitivesColorsEmotionalDanger600
        !enabled -> KozmosThemeTokens.primitivesColorsForeground500
        else -> MaterialTheme.colorScheme.onSurface
    }
    return SelectionFieldColors(
        background = if (!enabled || readOnly) KozmosThemeTokens.primitivesColorsBackground100 else KozmosThemeTokens.primitivesColorsBackground0,
        border = statusColor ?: KozmosThemeTokens.primitivesColorsForeground500,
        text = textColor,
        placeholder = if (!enabled) KozmosThemeTokens.primitivesColorsForeground500 else KozmosThemeTokens.primitivesColorsForeground400,
        label = statusColor ?: textColor,
        helper = when {
            !enabled -> KozmosThemeTokens.primitivesColorsForeground500
            statusColor != null -> statusColor
            else -> KozmosThemeTokens.primitivesColorsForeground500
        }
    )
}
